File: tests/omp/tasking10.cpp
#include <omp.h>
#include <stdio.h>
void work() {
    omp_lock_t lock;
    {
      for (int i = 0; i < 100; i++) {
        {
        // lock is shared by default in the task
                // Capture data for the following task
        // Task Scheduling Point 1
          { printf ("do work here\n"); }
        }
      }
    }
}

File: tests/omp/tasking10.ml
open Optitrust
open Target

let _ = Run.script_cpp (fun _ ->
  
  !! Omp.init_lock "lock" [tAfter; cVarDef "lock"];
  !! Omp.parallel [] [tBefore; cSeq ~args:[cFor "i"] ()];
  !! Omp.for_ [] [tBefore;cFor "i"];
  !! Omp.task [] [tFirst; cFor "i"; dBody];
  !! Omp.set_lock "lock" [tBefore; cSeq ~args:[cFun "printf"] ()];
  !! Omp.task [] [tBefore; cSeq ~args:[cFun "printf"] ()];
  !! Omp.unset_lock "lock" [tAfter; cSeq ~args:[cFun "printf"] ()];
  !! Omp.task [] [tFirst; cFunDef "work";dBody];
  !! Omp.destroy_lock "lock" [tLast; cFunDef "work"; dBody];
  

)
